code/modules/lavaland/caves_theme.dm

```diff
@@ -0,0 +1,198 @@
+			return
+
+	// there can be only one bubblegum, so don't waste spawns on it
+	if(ispath(mob_spawn, /mob/living/simple_animal/hostile/megafauna/bubblegum))
+		GLOB.megafauna_spawn_list.Remove(mob_spawn)
+
+	// same as above, we do not want multiple of these robots
+	if(ispath(mob_spawn, /mob/living/simple_animal/hostile/megafauna/ancient_robot))
+		GLOB.megafauna_spawn_list.Remove(mob_spawn)
+
+	new mob_spawn(T)
+	SSblackbox.record_feedback("tally", "lavaland_mob_spawns", 1, "[mob_spawn]")
+
+/proc/lavaland_caves_spawn_flora(turf/T)
+	var/flora_spawn = pickweight(GLOB.caves_default_flora_spawns)
+	for(var/obj/structure/flora/ash/F in range(4, T)) //Allows for growing patches, but not ridiculous stacks of flora
+		if(!istype(F, flora_spawn))
+			return
+	new flora_spawn(T)
+
+/datum/caves_theme
+	var/name = "Not Specified"
+
+	var/seed
+	var/perlin_accuracy = 5
+	var/perlin_stamp_size = 10
+	var/perlin_lower_range = 0
+	var/perlin_upper_range = 0.3
+
+/datum/caves_theme/New()
+	seed = rand(1, 999999)
+
+/datum/caves_theme/proc/setup()
+	var/result = rustg_dbp_generate("[seed]", "[perlin_accuracy]", "[perlin_stamp_size]", "[world.maxx]", "[perlin_lower_range]", "[perlin_upper_range]")
+	var/z = level_name_to_num(MINING)
+	for(var/turf/T in block(1, 1, z, world.maxx, world.maxy, z))
+		if(!istype(get_area(T), /area/lavaland/surface/outdoors/unexplored))
+			continue
+		if(!istype(T, /turf/simulated/mineral))
+			continue
+		var/c = result[world.maxx * (T.y - 1) + T.x]
+		if(c == "1")
+			T.ChangeTurf(/turf/simulated/floor/plating/asteroid/basalt/lava_land_surface)
+			on_change(T)
+
+/datum/caves_theme/proc/on_change(turf/T)
+	if(prob(2))
+		lavaland_caves_spawn_flora(T)
+	else if(prob(1))
+		lavaland_caves_spawn_mob(T)
+
+/datum/caves_theme/proc/safe_replace(turf/T)
+	if(!istype(get_area(T), /area/lavaland/surface/outdoors/unexplored))
+		return FALSE
+	if(istype(T, /turf/simulated/floor/chasm))
+		return FALSE
+	if(istype(T, /turf/simulated/floor/lava/lava_land_surface))
+		return FALSE
+	if(istype(T, /turf/simulated/floor/lava/mapping_lava))
+		return FALSE
+
+	return TRUE
+
+/datum/caves_theme/classic
+	name = "Classic Caves"
+
+/datum/caves_theme/burrows
+	name = "Blocked Burrows"
+	perlin_accuracy = 90
+	perlin_stamp_size = 7
+	perlin_lower_range = 0
+	perlin_upper_range = 0.3
+
+/datum/caves_theme/burrows/on_change(turf/T)
+	if(prob(9))
+		new /obj/structure/flora/ash/rock/style_random(T)
+	else if(prob(5))
+		lavaland_caves_spawn_flora(T)
+	else if(prob(1))
+		lavaland_caves_spawn_mob(T)
+
+/datum/caves_theme/deeprock/proc/maybe_make_room(turf/T)
+	if(rand(1, 150) != 1)
+		return
+
+	for(var/turf/oasis_centroid in oasis_centroids)
+		if(get_dist(T, oasis_centroid) < oasis_padding)
+			return
+
+	oasis_centroids |= T
+	var/new_scan_range = rand(4, 7)
+	var/tempradius = rand(10, 15)
+	var/probmodifer = 43 * tempradius //Yes this is a magic number, it is a magic number that works well.
```

Comment:
It's probably worth making this 43 into a define, for readability.

Reply:
Done.